src/main/flight/imu.c

```diff
@@ -377,14 +385,87 @@ static void imuMahonyAHRSupdate(float dt, const fpVector3_t * gyroBF, const fpVe
     /* Step 2: Roll and pitch correction -  use measured acceleration vector */
     if (accBF) {
         static const fpVector3_t vGravity = { .v = { 0.0f, 0.0f, 1.0f } };
-        fpVector3_t vEstGravity, vAcc, vErr;
 
-        // Calculate estimated gravity vector in body frame
-        quaternionRotateVector(&vEstGravity, &vGravity, &orientation);    // EF -> BF
+        static bool         vVelEF_initialized;
+        static fpVector3_t  vVelEF_prev;
+        static fpVector3_t  vVelEF_AccIntegal;      // Integrated accelerometer value in EF
+        static float        vVelEF_integralTime;
+        static fpVector3_t  vVelEF_errorVector;
+
+        fpVector3_t         vErr;
+
+        if (velEF && vVelEF_initialized) {
+            // We have valid velocity from GPS, use that to calculate correction vector
+            fpVector3_t vAccEF;
+
+            // Rotate ACC from BF to EF and accumulate
+            quaternionRotateVectorInv(&vAccEF, accBF, &orientation);
+            vectorAdd(&vVelEF_AccIntegal, &vVelEF_AccIntegal, &vAccEF);
+            vVelEF_integralTime += dt;
+
+            // Now if we got a GPS update we can calculate error vector according to
+            // the paper by Bill Premerlani (Roll-Pitch Gyro Drift Compensation)
+            if (velEFNew) {
+                fpVector3_t vTmp1, vTmp2;
+
+                // Calculate (V2 - V1) / (T2 - T1)
+                // Here T2-T1 = vVelEF_integralTime (accumulated time between two GPS updates, also time window we used to integrate accelerometer)
+                vectorSub(&vTmp1, &vVelEF_prev, velEF);
+                vectorScale(&vTmp1, &vTmp1, 1.0f / vVelEF_integralTime);
+
+                // Calculate Ge - (V2 - V1) / (T2 - T1)
+                vectorSub(&vTmp1, &vGravity, &vTmp1);
+                const float vTmp1Length = sqrtf(vectorNormSquared(&vTmp1));
+
+                if (vTmp1Length > 0.01f) {
+                    // Calculate acceleration by taking a derivative of acceleration integral
+                    // This effectively calculates average acceleration, but in a way that's more immune to jitter
+                    vectorScale(&vTmp2, &vVelEF_AccIntegal, 1.0f / vVelEF_integralTime);
```

Comment:
There is something wrong there, vTmp2 which should have average acceleration between GPS updates in cm/s/s seems like to be multiplied by a factor of 1000